Muehle_Algo.ipynb

```diff
@@ -0,0 +1,353 @@
+    "                [State[1][1][0],State[1][1][1],State[1][1][2],State[1][1][3],State[1][1][4],State[1][1][5],State[1][1][6],State[1][1][7]],\n",
+    "                [State[1][2][0],State[1][2][1],State[1][2][2],State[1][2][3],State[1][2][4],State[1][2][5],State[1][2][6],State[1][2][7]]\n",
+    "    ]]\n",
+    "\n",
+    "to_tupel(START_STATE)\n",
+    "\n",
+    "to_list(to_tupel(START_STATE))"
+   ]
+  },
+  {
+   "source": [
+    "Die Funktion `memoize(f)` aus der \"Wissensbasierte Systeme\"-Vorlesung nimmt ein Argument f und gibt eine Version dieser Funktion zurück, die alle Ergebnisse der Funktion f zwischenspeichert."
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def memoize(f):\n",
+    "    global Cache\n",
+    "    \n",
+    "    def f_memoized(*args):\n",
+    "        key = (args[0], args[1])\n",
+    "        if key in Cache:\n",
+    "            return Cache[key]\n",
+    "        result = f(*args)\n",
+    "        Cache[key] = result\n",
+    "        return result\n",
+    "    \n",
+    "    return f_memoized"
+   ]
+  },
+  {
+   "source": [
+    "## Minimax-Algorithmus"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import random\n",
+    "random.seed(1)"
+   ]
+  },
+  {
+   "source": [
+    "`value_minimax(State, player)`"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@memoize\n",
+    "def value_minimax(State, player, depth):\n",
+    "    if finished(to_list(State)):\n",
+    "        return utility(to_list(State), player)\n",
+    "    if depth == 0:\n",
+    "        return heuristic(State, player)\n",
+    "    o = opponent(player)\n",
+    "    depth = depth - 1\n",
+    "    return max([ -value_minimax(to_tupel(ns), o, depth) for ns in next_states(to_list(State), player) ])"
+   ]
+  },
+  {
+   "source": [
+    "`best_move_minimax(State, player)`"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def best_move_minimax(State, player, depth):\n",
+    "    NS        = next_states(State, player)\n",
+    "    bestVal   = value_minimax(to_tupel(State), player, depth)\n",
+    "    BestMoves = [s for s in NS if -value_minimax(to_tupel(s), opponent(player), depth) == bestVal]\n",
+    "    BestState = random.choice(BestMoves)\n",
+    "    return bestVal, BestState"
+   ]
+  },
+  {
+   "source": [
+    "Die Funktion `minimax(State, Player)` wurde erstellt, um die Funktion `best_move_minimax(State, player)` nach Außen eindeutiger von `alpha_beta_pruning(State, Player)` abzugrenzen."
```

Comment:
```suggestion
    "Die Funktion `minimax()` wurde erstellt, um die Funktion `best_move_minimax()` nach außen hin eindeutiger von `alpha_beta_pruning()` abzugrenzen."
```

Reply:
Vielleicht sollten wir generell mal drüber reden. Ich habe bei den Dokumentationen meistens die Parameter einer Funktion weggelassen, damit es nicht zu lang wird und wir sagen ja danach eh, wie die Funktion aufgerufen wird. Können wir ja nochmal drüber reden

Reply:
werd ich in Issue #84 bearbeiten